modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/MajorityPeerSelect.scala

```diff
@@ -50,28 +60,34 @@ object MajorityPeerSelect {
 
     def filterPeerList(peers: NonEmptyList[Peer]): F[NonEmptyList[Peer]] =
       peers
-        .traverse(snapshotClient.getLatestOrdinal(_))
-        .map {
-          _.groupBy(identity).maxBy { case (_, ordinals) => ordinals.size }
+        .parTraverseN(maxConcurrentPeerInquiries)(snapshotClient.getLatestOrdinal(_))
+        .map(_.groupBy(identity).maxBy { case (_, ordinals) => ordinals.size })
+        .map { case (majorityOrdinal, _) => majorityOrdinal }
+        .flatMap { majorityOrdinal =>
+          peers
+            .parTraverseN(maxConcurrentPeerInquiries)(peer => getPeerHash(peer, majorityOrdinal).attempt)
         }
-        .flatMap {
-          case (majorityOrdinal, _) =>
-            peers.traverse(snapshotClient.get(majorityOrdinal).run(_).flatMap(_.toHashed.map(_.hash)))
+        .flatMap { maybePeerSnapshots =>
+          MonadThrow[F].fromOption(
+            maybePeerSnapshots.collect { case Right(peerSnapshot) => peerSnapshot }.toNel,
+            NoPeersToSelect
+          )
         }
-        .map(_.zip(peers))
-        .map(_.groupMap { case (hash, _) => hash } { case (_, ps) => ps })
-        .map(_.maxBy { case (_, peers) => peers.size })
-        .map { case (_, peerCandidates) => peerCandidates }
+        .map(_.groupMap { case (_, hash) => hash } { case (peer, _) => peer })
+        .map(_.values.toList.sortBy(-_.size))
+        .map(_.head)
 
     def getPeerSublist(peers: Set[Peer]): F[List[Peer]] = {
-      val maxSublistPercent = 0.25
-      val maxSublistSize = Math.max((peers.size * maxSublistPercent).toInt, 1)
+      val maxSampleSize = minPeerSelectionSize + (peers.size * peerSampleRatio).toInt
 
-      Random[F].nextIntBounded(maxSublistSize).flatMap { peerCount =>
+      Random[F].betweenInt(minPeerSelectionSize, maxSampleSize + 1).flatMap { peerCount =>
```

Comment:
why `+1` if using `betweenInt`?

Reply:
Because if `(peers.size * peerSampleRatio).toInt` is 0, then `maxSampleSize` == `minPeerSelectionSize`. @sdavidp  and I were unsure what the behavior would be if the upper and lower bounds were the same value. `+1` ensures that the upper bound is always greater than the lower bound.

Reply:
I'd rather think about the algorithm and then implement it, instead of thinking how specific implementation works and craft the algorithm to match the implementation.

What's the requirement of this ticket? You want to select some subset of peers for majority selection, right?
IMO, `minPeerSelectionSize` says that it can be the lower bound. the upper bound is max of 25% peers or minPeerSelectionSize if I understood it correctly. so, if we have less peers than minPeerSelectionSize, we should select all the peers for majority selection. do we want to randomly choose number of peers instead of 25% of the cluster? if so, it can look like:

peers > minPeerSelectionSize => take random number between minPeerSelectionSize and 25% of peers
peers <= minPeerSelectionSize => take all peers

Reply:
Implemented the algorithm suggested by @marcinwadon. `PeerSelect` will either select the the discovered `Peer`s up to `minSampleSize` (20) or 25% of the discovered `Peer`s, whichever is higher.